Implement #3

TheEpicBlock opened the pull request:
This implements the test vectors as outlined in #3 for all implementations except for dlang. I copied over all testcases I could find. All testcases succeed except for `a-a < a`, which was included in go's testcases but fails in all other implementations, I'm not sure what the correct behaviour is here.

Instead of the large int test being optional, it's in a separate file and implementations chose whether to include that file in their tests. This means that the large int test is fully enforced for those languages (so if it regresses you would notice).

- [x] I own the full rights to my contribution, and agree to release it under the terms of the Creative Commons Zero Public Domain Dedication

TheEpicBlock committed:
Implement #3

Infinidoge commented on rust/src/lib.rs:
```diff
@@ -161,70 +161,81 @@ impl Ord for FlexVer<'_> {
-                Equal => Equal,
-            }
+            FlexVer("1.1.0"),
+            FlexVer("1.1.0").clamp(FlexVer("1.0.0"), FlexVer("1.2.0"))
```
Nitpick: I think `test_clamp` should be below `test_min`/`test_max`
It just feels like the right place.

Infinidoge commented on js/test.js:
```diff
@@ -1,43 +1,43 @@
+				return runTest(lefthand, righthand, ordening)
+			})
+	})
+})
```
```suggestion
})

```
Nitpick: Should have a newline at the end of the file

Infinidoge commented on java/src/test/java/com/unascribed/flexver/Ordering.java:
```diff
@@ -0,0 +1,52 @@
+
+        throw new IllegalStateException();
+    }
+}
```
```suggestion
}

```
Nitpick: Should have a newline at the end of the file

exaskye commented on java/src/test/java/com/unascribed/flexver/Ordering.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.NoSuchElementException;
+
+public enum Ordering {
+    Less("<"),
+    Equal("="),
+    Greater(">");
```
Conventionally, enum constants (all constants, in fact) are SCREAMING_SNAKE_CASE.

```suggestion
    LESS("<"),
    EQUAL("="),
    GREATER(">");
```
(Suggestion must be applied via IDE refactor, not in GitHub web.)

TheEpicBlock replied:
That's pretty embarrassing. Guess I've been doing too much Rust lately

exaskye commented on java/src/test/java/com/unascribed/flexver/Ordering.java:
```diff
@@ -0,0 +1,52 @@
+        if (i == 0) return Ordering.Equal;
+        if (i > 0) return Ordering.Greater;
+
+        throw new IllegalStateException();
```
Should be `AssertionError`

exaskye commented on java/src/test/java/com/unascribed/flexver/Ordering.java:
```diff
@@ -0,0 +1,52 @@
+        }
+    }
+
+    public static Ordering fromStr(String str) {
```
from or fromString

unascribed commented on test/test_vectors.txt:
```diff
@@ -0,0 +1,59 @@
+
+# Pre-releases can have multiple dashes (should not be split)
+# Reasoning for test data: "p-a!" > "p-a-" (correct); "p-a!" < "p-a t-" (what happens if every dash creates a new component)
+-a- > -a!
```
Didn't realize the Go impl had some new good tests in it like this one.

exaskye commented on test/test_vectors.txt:
```diff
@@ -0,0 +1,59 @@
+ < 1
+
+# Check boundary between textual and prerelease
+a-a < a
```
I believe this describes the correct behavior. Why are all impls but Go failing this...?

exaskye replied:
@Infinidoge seems to have figured it out; in Go, `-` is treated specially and splits out a component, but all the other implementations look for *textual components* that are *already split off* that start with `-`.

Checking the spec, that *is* correct behavior, but I don't know if that's the *intended* behavior...

unascribed commented on java/src/test/java/com/unascribed/flexver/FlexVerBasicTest.java:
```diff
@@ -1,92 +1,49 @@
-		return i < 0 ? -1 : i > 0 ? 1 : 0;
-	}
+	@ParameterizedTest(name = "{0} {2} {1}")
+	@MethodSource("getEqualityTests")
```
Well, I learned something new about JUnit today.

exaskye commented on java/build.gradle:
```diff
@@ -8,6 +8,20 @@ sourceCompatibility = targetCompatibility = 8
+dependencies {
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.2'
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.2'
+	implementation 'org.junit.jupiter:junit-jupiter-params:5.9.2'
```
This will pull in this dep transitively on Maven! Needs to be `compileOnly` or `testImplementation`.

exaskye reviewed (commented):
Oh, I wasn't expecting someone to PR this; that was a TODO for myself, heh. Thanks!

I can only satisfactorily review the Java and JS implementations; going to need codeowner reviews for the others.

JS: :heavy_check_mark: 
Java: :heavy_minus_sign: (but just nits)

Infinidoge reviewed (approved):
The Rust implementation looks good to me!
(Though I'm not sure about the `a-a < a` situation either, will need Una's input on that)

Had a few nitpicks, mainly about newlines at the ends of files, but other that the PR looks solid.

I've only personally tested the Rust implementation.

exaskye committed:
Update js/test.js

Co-authored-by: Infinidoge <infinidoge@inx.moe>

exaskye committed:
Update java/src/test/java/com/unascribed/flexver/Ordering.java

Co-authored-by: Infinidoge <infinidoge@inx.moe>

TheEpicBlock committed:
Code style improvements

TheEpicBlock committed:
Fix maven transitivety

comp500 commented on go/flexver/main_test.go:
```diff
@@ -9,219 +9,181 @@
-		return ">"
-	} else {
-		return "!"
+func RunCompare(t *testing.T, lefthand string, righthand string, ordening int) {
```
Spelling mistake in `ordening` (also for `RunLess`/`RunEqual`)

comp500 commented on go/flexver/main_test.go:
```diff
@@ -9,219 +9,181 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer file.Close()
```
This won't close each individual test data file until the whole function returns; the for loop contents could be refactored out into another function (or inline closure)

comp500 commented on go/flexver/main_test.go:
```diff
@@ -9,219 +9,181 @@
-			}
-		})
+func TestInvalid(t *testing.T) {
+	_, err := CompareError("\xc3\x28", "")
```
This no longer tests whether the panicking variants panic on invalid input; `defer func() { _ = recover() }()` can be used to catch a panic as done in the previous implementation.

comp500 reviewed (changes requested):
The Go implementation looks good, just got a few suggestions.

TheEpicBlock committed:
Fix spelling of ordering

TheEpicBlock committed:
Refactor testfile processing into seperate function

This ensures that the defer statement that's closing the testfile runs for each testfile individually, instead of all at once

TheEpicBlock committed:
Test for panics in go
